File: hostel-management-frontend/src/services/hostelService.ts
import api from "./api";
import { Hostel } from "../types";

export const hostelService = {
  async getAll(): Promise<Hostel[]> {
    const response = await api.get<Hostel[]>("/hostels");
    return response.data;
  },

  async addHostel(name: string): Promise<Hostel> {
    const response = await api.post<Hostel>("/hostels", { name });
    return response.data;
  },

  async addBhavana(hostelId: string, name: string): Promise<Hostel> {
    const response = await api.post<Hostel>(`/hostels/${hostelId}/bhavanAs`, { name });
    return response.data;
  },

  async addBlock(hostelId: string, bhavanaId: string): Promise<Hostel> {
    const response = await api.post<Hostel>(`/hostels/${hostelId}/blocks`, { bhavanaId });
    return response.data;
  },

  async addFloor(hostelId: string, bhavanaId: string, blockId: string, floorNumber: string): Promise<Hostel> {
    const response = await api.post<Hostel>(`/hostels/${hostelId}/floors`, {
      bhavanaId,
      blockId,
      floorNumber
    });
    return response.data;
  },

  async addRoom(
    hostelId: string,
    bhavanaId: string,
    blockId: string,
    floorId: string,
    roomNumber: string,
    type: number
  ): Promise<Hostel> {
    const response = await api.post<Hostel>(`/hostels/${hostelId}/rooms`, {
      bhavanaId,
      blockId,
      floorId,
      roomNumber,
      type
    });
    return response.data;
  },

  async addBed(
    hostelId: string,
    bhavanaId: string,
    blockId: string,
    floorId: string,
    roomId: string
  ): Promise<Hostel> {
    const response = await api.post<Hostel>(`/hostels/${hostelId}/beds`, {
      bhavanaId,
      blockId,
      floorId,
      roomId
    });
    return response.data;
  },

  async deleteNode(hostelId: string, level: string, nodeId: string): Promise<Hostel> {
    const response = await api.delete<Hostel>(`/hostels/${hostelId}/node`, {
      data: { level, nodeId }
    });
    return response.data;
  },

  async renameNode(hostelId: string, level: string, nodeId: string, newName: string): Promise<Hostel> {
    const response = await api.put<Hostel>(`/hostels/${hostelId}/node`, {
      level,
      nodeId,
      newName
    });
    return response.data;
  }
};

File: hostel-management-backend/src/main/java/com/iiitb/hostel/service/HostelService.java
package com.iiitb.hostel.service;

import java.util.List;

import com.iiitb.hostel.dto.AddBedRequest;
import com.iiitb.hostel.dto.AddBlockRequest;
import com.iiitb.hostel.dto.AddBhavanaRequest;
import com.iiitb.hostel.dto.AddFloorRequest;
import com.iiitb.hostel.dto.AddHostelRequest;
import com.iiitb.hostel.dto.AddRoomRequest;
import com.iiitb.hostel.dto.RenameNodeRequest;
import com.iiitb.hostel.model.Hostel;

public interface HostelService {
    List<Hostel> getAll();

    Hostel addHostel(AddHostelRequest request);

    Hostel addBhavana(String hostelId, AddBhavanaRequest request);

    Hostel addBlock(String hostelId, AddBlockRequest request);

    Hostel addFloor(String hostelId, AddFloorRequest request);

    Hostel addRoom(String hostelId, AddRoomRequest request);

    Hostel addBed(String hostelId, AddBedRequest request);

    Hostel deleteNode(String hostelId, String level, String nodeId);

    Hostel renameNode(String hostelId, RenameNodeRequest request);

    List<Hostel.Bhavana> getBhavanAs();

    List<Hostel.Block> getBlocks(String bhavanaId);

    List<Hostel.Floor> getFloors(String blockId);

    List<Integer> getRoomTypes();
}
